device trust marshalers

flyinghermit opened the pull request:
Include device in `tctl get all` command. 
Fixes: https://github.com/gravitational/teleport/issues/23775

Service Marshalers included in the enterprise PR: https://github.com/gravitational/teleport.e/pull/1563

Bootstrapping device will be supported in the subsequent PR. See issue: https://github.com/gravitational/teleport/issues/24667

flyinghermit committed:
feat: device resource in tctl get all

flyinghermit committed:
check for device resource but ignore in favour of enterprise resource migration

flyinghermit committed:
Merge branch 'master' into sshah/tctl-get-all-devices

flyinghermit committed:
Merge branch 'master' into sshah/tctl-get-all-devices

flyinghermit committed:
add firstStart indicator in auth service config. Tests

flyinghermit committed:
Merge with master

flyinghermit committed:
Merge branch 'master' into sshah/tctl-get-all-devices

flyinghermit committed:
remove device bootstrap from this PR

sshahcodes commented on lib/auth/init_test.go:
```diff
@@ -1368,3 +1368,52 @@ func TestMigrateDatabaseCA(t *testing.T) {
+  update_time: "2023-04-20T09:03:53.427809Z"
+version: v1`
+
+func TestFlagEResources(t *testing.T) {
```
Renamed in: https://github.com/gravitational/teleport/pull/24963/commits/79c0796833d4f344d1633935557589111b9535a1

sshahcodes commented on lib/auth/init_test.go:
```diff
@@ -1368,3 +1368,52 @@ func TestMigrateDatabaseCA(t *testing.T) {
 	require.Equal(t, hostCA.Spec.ActiveKeys.TLS[0].Key, dbCAs[0].GetActiveKeys().TLS[0].Key)
 }
+
+const deviceYAML = `kind: device
```
Done in https://github.com/gravitational/teleport/pull/24963/commits/79c0796833d4f344d1633935557589111b9535a1

sshahcodes commented on lib/services/local/resource.go:
```diff
@@ -101,6 +101,10 @@ func itemsFromResource(resource types.Resource) ([]backend.Item, error) {
+	case *types.DeviceV1:
+		// devices should be ignored here and instead restored
+		// using enterprise resource bootstrapper.
+		return nil, nil
```
Noted. Thanks for the context. Fixed in https://github.com/gravitational/teleport/pull/24963/commits/79c0796833d4f344d1633935557589111b9535a1 by returning empty slice: `return []backend.Item{}, nil`.

However, this has to be included in this PR otherwise if this PR is merge and until "bootstrap" is supported `tctl get all > backup.yml && teleport start --bootstrap=backup.yml` will not not work as we return `NotImplemented` error below this block.

flyinghermit committed:
Merge branch 'master' into sshah/tctl-get-all-devices

fspmarshall commented on api/types/device.go:
```diff
@@ -90,6 +90,15 @@ func UnmarshalDevice(raw []byte) (*DeviceV1, error) {
 
+// MarshalDevice marshals a DeviceV1 resource.
+func MarshalDevice(dev *DeviceV1) ([]byte, error) {
+	devBytes, err := json.Marshal(dev)
```
nit: we typically prefer using `utils.FastMarshal` over `json.Marshal`.  Using that would require moving this function to somewhere in `lib` since you can't import `lib/utils` from the `api/*` family of packages, but if there isn't a particular need for this function to live here, that might be worth doing.  Most typed marshal/unmarshal functions are defined in `lib/services` anyhow.

sshahcodes replied:
I decided to keep it as is because the way device services are currently setup, there's nothing to be added to `lib/services` except for these marshal/unmarshal func. For other services in `lib/services`, we have interface defined for them and then the marshalers works with the interface. We already have two types for Devices (`Device` and `DeviceV1`) and adding device marshalers to `lib/services` in similar way will (IMO) only complicate things (more scattered code for Device trust).

sshahcodes replied:
Moved device marshalers to `lib/services` https://github.com/gravitational/teleport/pull/24963/commits/6c4b9ad9b66c1b9411c4f7e917d25b6c0b623361

flyinghermit committed:
Merge branch 'master' into sshah/tctl-get-all-devices

sshahcodes commented on lib/auth/init.go:
```diff
@@ -705,6 +707,11 @@ func checkResourceConsistency(ctx context.Context, keyStore *keystore.Manager, c
 			if err := checkRoleFeatureSupport(r); err != nil {
 				return trace.Wrap(err)
 			}
+		case *types.DeviceV1:
```
Sorry I missed this comment! The OIDC and SAML have the same mechanism to register resources except they simply ignores this check (which I think we should). We have similar check for `types.Role:` which checks enterprise build for `SearchAsRoles`. For example, in `init.go`:
https://github.com/gravitational/teleport/blob/7143be2d042abf9c7bb4bcdbbe906ecc5def402f/lib/auth/init.go#L711 uses `checkRoleFeatureSupport` to check for enterprise build. https://github.com/gravitational/teleport/blob/7143be2d042abf9c7bb4bcdbbe906ecc5def402f/lib/auth/auth_with_roles.go#L3657

sshahcodes replied:
my bad, I think you meant to refer to this resource marshaler from e [modules](https://github.com/gravitational/teleport.e/blob/master/tool/modules/resources.go)! let me check that.

sshahcodes replied:
@rosstimothy I added resource marshaler to e resource marshaler - https://github.com/gravitational/teleport.e/pull/1563/files

Can you please review this again?

flyinghermit committed:
Merge branch 'master' into sshah/tctl-get-all-devices

flyinghermit committed:
multiple updates:
- remove resource marshaler, this will be added to e repo instead
- remove device resource checks in oss (not needed as resource marshaler now added to e). tests removed

flyinghermit committed:
Merge branch 'sshah/tctl-get-all-devices' of github.com:gravitational/teleport into sshah/tctl-get-all-devices

flyinghermit committed:
move device marshalers to service package

flyinghermit committed:
resolve conflict with master, updated marshalers to use utils.Fastmarshal

flyinghermit committed:
run fix-imports, fix test

flyinghermit committed:
remove device case from itemsFromResource

sshahcodes commented on lib/services/local/resource.go:
```diff
@@ -101,6 +101,10 @@ func itemsFromResource(resource types.Resource) ([]backend.Item, error) {
+	case *types.DeviceV1:
+		// Return empty slice here as bootstrapping devices will be supported with
+		// enterprise resource bootstrapper.
+		return []backend.Item{}, nil
```
Removed in https://github.com/gravitational/teleport/pull/24963/commits/cf26a725300e6d7eb539ed8301d923c3378a0ed3.

rosstimothy reviewed (approved):
LGTM. Though I would maybe update the PR title/description and the commit message if you squash to a single commit before merging to better reflect that this is now just moving marshaling of devices into `lib/services` and that the actual support for `tctl get all` exists entirely in your other PR in `e`.
